pkg/morph/deploy/notary.go

```diff
@@ -0,0 +1,880 @@
+					zap.Uint32("current height", cur), zap.Uint32("retry after height", designateRoleTxValidUntilBlock))
+				return
+			}
+
+			prm.logger.Info("previously sent transaction designating Notary role to the committee expired without side-effect, will recreate")
+			generateAndShareTxData(true)
+			return
+		}
+
+		if !txFullySigned {
+			prm.logger.Info("finalizing the transaction designating Notary role to the committee...")
+
+			initialLen := len(tx.Scripts[1].InvocationScript)
+			var extraLen int
+
+			for _, sig := range mCommitteeIndexToSignature {
+				extraLen += 1 + 1 + len(sig) // opcode + length + value
+			}
+
+			tx.Scripts[1].InvocationScript = append(tx.Scripts[1].InvocationScript,
+				make([]byte, extraLen)...)
+			buf := tx.Scripts[1].InvocationScript[initialLen:]
+
+			for _, sig := range mCommitteeIndexToSignature {
+				buf[0] = byte(opcode.PUSHDATA1)
+				buf[1] = byte(len(sig))
+				buf = buf[2:]
+				buf = buf[copy(buf, sig):]
+			}
+
+			txFullySigned = true
+		}
+
+		prm.logger.Info("sending the transaction designating Notary role to the committee...")
+
+		_, vub, err := _actor.Send(tx)
+		if err != nil {
+			designateRoleTxValidUntilBlock = 0
+			switch {
+			default:
+				prm.logger.Error("failed to send transaction designating Notary role to the committee, will try again later",
+					zap.Error(err))
+			case isErrNotEnoughGAS(err):
+				prm.logger.Info("not enough GAS for transaction designating Notary role to the committee, will try again later")
+			case isErrInvalidTransaction(err):
+				prm.logger.Warn("composed transaction designating Notary role to the committee is invalid and can't be sent, will recreate",
+					zap.Error(err))
+				generateAndShareTxData(true)
+			}
+			return
+		}
+
+		designateRoleTxValidUntilBlock = vub
+
+		prm.logger.Info("transaction designating Notary role to the committee has been successfully sent, will wait for the outcome")
+	}, nil
+}
+
+// initDesignateNotaryRoleAsSignerTick returns a function that preserves context
+// of the Notary role designation to the multi-acc committee between calls. The
+// operation is performed by the non-leading committee member which is assigned to
+// sign transaction submitted by the leader.
+func initDesignateNotaryRoleAsSignerTick(prm enableNotaryPrm, monitor *blockchainMonitor) (func(), error) {
+	committeeMultiSigM := smartcontract.GetMajorityHonestNodeCount(len(prm.committee))
+	committeeMultiSigAcc := wallet.NewAccountFromPrivateKey(prm.localAcc.PrivateKey())
+
+	err := committeeMultiSigAcc.ConvertMultisig(committeeMultiSigM, prm.committee)
+	if err != nil {
+		return nil, fmt.Errorf("compose committee multi-signature account: %w", err)
+	}
+
+	_actor, err := actor.NewSimple(prm.blockchain, prm.localAcc)
+	if err != nil {
+		return nil, fmt.Errorf("init transaction sender from local account: %w", err)
+	}
+
+	committeeSigners := []actor.SignerAccount{
```

Comment:
@cthulhu-rider, @roman-khimov can you, please, explain me in short why do we have two signers here?

Reply:
Because the first one pays for the transaction, while the second one witnesses a role change.